### DISCUSSION TITLE
Client not allowed for direct access grants

### DISCUSSION THREAD

**Discussion by Nickk4** [keycloak/keycloak#22472] [Q&A] [answered]
I'm using Keycloak v21.0.1 and authentication via my application works. But when I try to get an access token via Postman it fails with the error "Client not allowed for direct access grants".

![image](https://github.com/keycloak/keycloak/assets/10728166/2f24b5fe-7b12-4b21-8e12-302c4af167a5)

Direct access grants is enabled for the client:

![image](https://github.com/keycloak/keycloak/assets/10728166/033b00b2-3227-47b5-9371-41fc24f9186d)

And I also read some had the problem because they were working with "temporary passwords". But also that is not the case for me.

What could be the cause of the error "Client not allowed for direct access grants"?

**Comment by justin-tay** [comment]
From what I can tell you should only get that particular error if the client didn't have `Direct access grants` enabled so it is rather strange you are getting it when it is enabled.

https://github.com/keycloak/keycloak/blob/48a4102c9700ed50034f2560078b7b9d0cd1e952/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L649-L652

If the user had temporary passwords it would be a different error which is `Account is not fully set up`. 

https://github.com/keycloak/keycloak/blob/48a4102c9700ed50034f2560078b7b9d0cd1e952/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L696-L702

Maybe you could try again with a test realm and client?

![direct](https://github.com/keycloak/keycloak/assets/49700559/c5f56e27-d5fb-4ac3-999d-f6ba1496ffba)

**Comment by Nickk4** [comment, marked as answer]
Thanks, got it solved. The cause turned out to be a caching issue: when changing settings in KeyCloak, it only emptied the cache of one pod and not all of them.
